### ISSUE TITLE
[EXWM] Not running under X environment message at login

### ISSUE THREAD

**Issue by doolio** [ch11ng/exwm#703]
Firstly, great package. 

Running exwm-0.23. The last three lines of my ~/.xinitrc are as follows:

```
picom -b
## `exwm-enable' has to be called before the frame is spawned.
emacs --daemon --eval "(require 'exwm)" -f exwm-enable
exec emacsclient -c
```
However, I get the following message at login.

> [EXWM] Not running under X environment

Seems odd. Is this an error in how I start exwm?

**Comment by QiangF** [comment]
Why not try the provided configuration first to see if the problem exist in other places?
https://github.com/ch11ng/exwm/wiki/Configuration-Example

**Comment by ch11ng** [comment]
What's the result of `(framep (selected-frame))` for the first visible frame?

**Comment by doolio** [comment]
> What's the result of (framep (selected-frame)) for the first visible frame?

x

At least I believe that is the value of the first visible frame. That is as soon as I login into a TTY session the screen flashes light grey momentarily then Emacs opens maximized due to my ~/Xresources configuration so the frame border is not visible. I presume this is the first visible frame. I don't open a new frame in any case. Is my ~/.xinitrc configuration not creating the first visible frame?

When I evaluate that sexp the value is x as one might expect despite the message that EXWM is not running under an X environment. Before opening this issue I had searched past issues and the only one that raised the same message was #472 but I wasn't clear if this issue was related  to that one.

Thanks for your time.

**Comment by ch11ng** [comment]
The result of that requested form is okay.  The first frame is created as a result of `emacsclient -c` and it should be the only visible one.  What if you call `exwm-init` manually? You may want to enable `exwm-debug` from the very beginning and the debug log containing `exwm-init` should indicate which frames EXWM has ever tried.

**Comment by doolio** [comment]
The `exwm-debug` log can be found [here](https://pastebin.com/dR5Ek2Fx). Not sure what I should look for though. It seems only one frame is ever referenced which I guess is expected. When should I call `exwm-init` manually? Is that not called automatically by the lines in my ~/.xinitrc ? Thanks for your time.

**Comment by ch11ng** [comment]
From the debug logs it seems EXWM was started.  Perhaps the message was a result of your creating other TUI frame?

> When should I call exwm-init manually? Is that not called automatically by the lines in my ~/.xinitrc ?

In any graphical frame.  It's just a test and not intended to be called by users.  Now that the debug logs revealed it was invoked there's no need for the test any more.

**Comment by doolio** [comment]
> From the debug logs it seems EXWM was started. Perhaps the message was a result of your creating other TUI frame?

No, I don't create another frame (TUI or otherwise) at least not intentionally. I will continue to investigate and report back if I find the cause of the issue. Thanks for your time.

**Comment by doolio** [comment]
Playing some more with this. If I include an `&` to "force" the Emacs daemon into the background as follows an emacsclient frame is launched whereas it was not before which I presume was just a standard frame. In this emacsclient frame the message no longer appears. Does this help to understand why I saw the message before?

```
picom -b
## `exwm-enable' has to be called before the frame is spawned.
emacs --daemon --eval "(require 'exwm)" -f exwm-enable &
exec emacsclient -c
```

However, by doing this the daemon procss is classed as a defunct child process of the emacsclient. Is this what you would expect? In addition, the `exwm-init-hook`s in my config given in #704 get set but aren't enabled, nor does my theme get loaded. I don't understand this. I've been able to load my theme by adding it to an `after-make-frame-functions` hook after reading about a similar issue on SO but I'm not sure if that is the correct approach.

Also, just to clarify the user guide states we can use:

```
emacs --daemon -f exwm-enable
exec emacsclient -c
```

By which I presume one is expected to `(require 'exwm)` in their `init.el`. Perhaps that should be more emphasised in the user guide or that one should use one of the following lines in their `.xinitrc` if they want to avoid using `(require 'exwm)` in their `init.el`.

```
emacs --daemon --eval "(require 'exwm)" -f exwm-enable
emacs --daemon --eval "(progn (require 'exwm) (exwm-enable))"
```

**Comment by medranocalvo** [comment]
The reason you saw that message before is the following:

1. `exwm-enable` add makes sure `exwm-init` runs at most once on `window-setup-hook` or `after-make-frame-functions`.  That is, whenever a frame is created, `exwm-init` runs.
2. Emacs does not like running without frames, so the `--daemon` creates an initial frame "F1".
3. That initial frame is not an X11 frame.
4. `exwm-init` runs when that frame is created, due to the hooks above.
5. `exwm-init` informs that the created frame is not an X11 frame (implied: cannot initialize EXWM).

That message is (probably) helpful for people wondering "why doesn't EXWM start when I eval `exwm-init`?".

It is confusing in your case, @doolio.

An improvement would be to clarify the message:

> [EXWM] Window system of \<frame F1> is not an X; waiting for an X frame...

Not great, but looks like an informational message instead of an error.

A better solution would be, in my opinion, to divide `exwm-init` in two functions: one public, intended to be called by users, and another internal to be used by the former and `exwm-enable`.  The public `exwm-init` could throw errors relevant for users invoking it interactively, while the internal could print info messages relevant to `exwm-enable`.

**Comment by doolio** [comment]
I see. Thank you for explaining. I would welcome your improved message and/or your proposed solution. And thanks for taking the time.

**Comment by ch11ng** [comment]
Or maybe we just detect if `exwm-init` was called interactively with `called-interactively-p` and print different messages the users.

**Comment by zackteo** [comment]
Am getting this error after installing emacs-native-comp, are EXWM and gccemacs not compatible? :o

**Comment by zackteo** [comment]
I realise it was because I was using the pgtk-nativecomp branch. pgtk =/= x11 🤦‍♂️
